### ISSUE TITLE
external manipulate array vs ractive.update

### ISSUE THREAD

**Issue by enneid** [ractivejs/ractive#3277]
## Description:
Please see link. I manipulate array of elements (which are wrapped by adaptor) by my self, and only trigger ractive.update on  elements 
If I remove element, then push  another, and change its attribute, last element will be nullify.

After some digging in code i found that there remains newWrapperValue from removed element set on undefined (but key still remains) and that cause problem

## Versions affected:
1.0.0, edge
## Reproduction:
 [link](https://ractive.js.org/playground/index.html?env=docs&hidemenu#N4IgFiBcoE5SBTAJgcwSAvgGhAF3gDxICWAbgATFIC8AOngIYxq70B8BA9CaW7QHYCCAZwDGMYgAdclGvVwIAtpIA2DBfXK4AnpIR08CAB65OMBqNxkE7AXZnlgwAMTEFi4Rgz3y5WricGLwAyfgAjYUk-QQcnTjclT29+LjEJaTYQHGF4UiZKd2Fyal8AbWAGSD9DYVwARnosclEq+lEGzCwK1prcACZG5p7RAc7HSuqFWoBmQZbq0VnMAF0BACULK1IEADoGJAZpAHsYYR2AQXOD45hi32ABX18AM2IVBRgq54BXfktiI78AAU5DyKm+CCaAGsELp1GAmuZ-ttyABKRyPJ6+USA4RHFS7FRHFBAsEQppkhCUfi1Bh-BBHZ7kADyYQAVghLORgsFyED6AxNMR+KCGOCEOjUZisTAELhvjARZTqbT6YyWezOTIeXyBUKlWKIVKRU9sNL-AB3cySL6-f6AkFIraQ8hHdnQ2GSeFNSSy15GBC3dEPE0yuUKkUhrH+YRy21-KwO31HPQwHQUw0SjGhrFutmlZOpnTLO6U6WmrDl3yy2O4eP24HBqtPWXyxXkZ5i2MAbmbvk4nHIABF1AxyLWPh2TuQAArMgBSzPIMLhuDAO2bZpzvgUTCQRwt-Hricb2ax0e3TwH5AAwkckFTjJzvgokK6RbuYPvDxvL1v-ABuAsMeAKNlG57VuG7Z5pu5bJP4yTwfweS3E61h3PwCAWuQGzIggQIhv4uBMMB5AAOTEcwcpkZWJruKo6gIFUZHOPRagKDRdgmteVxvmuVJkQkihkeQXppsQYrkGAgY2CaYlWF2VTAFuvgHMRSlCcIVSaSp5DXsKqqiI+2z8DIYB0kgBKnJi+yHHW5ClPQlzXLgJz0KsIoCBgqK9oI-ADrKihHKQwooOQai1OQCAEooCCmeQFnkN8khqaFAiaTswhgMQzy4ECxoCDiNL4oSxJAppBX8Gh2w7Mlan4fQmn0JVA6SN8WWheQmHYdFShxTIiV1eooVNLUbwqFF2wSGunXEEUYoTRIKBgKw-AZW1WUERM8gILUAAs9DeVx1W7ENCj8iATUgMaRV4gSOxEiSFUCAOZ2AmFgJUuouASGEL5UsKXVYVFMX9Ql-BvmdnUzfNuA-cQf0aGthSlH0yyOSAgogCWJT0Io2hA9hwjJYGorivQx2bNYtUpYxF0ZX0ezNVxA7nM8pNnVSEUyL1sXxeqQmUEU-DfCoKiFbiJUPWVz38FkICiPAmBAA)

**Comment by evs-chris** [comment]
It will take me a bit to get familiar with adaptors again, but out of curiosity, why are you using an adaptor if you're also going to call `update`? Theoretically, if you are adapting the value, you shouldn't also have to call `update` on the keypath as the adaptor is supposed to handle that.

**Comment by enneid** [comment]
this case was my test about efficiency of using ractive.set and ractive.update. sometimes i have massive changes in data, and I realized that faster is to first change my model, and then call for examaple parent update.  In other case you made  multiple ractive.set,  which multiple modify DOM, calls observers ect.

**Comment by evs-chris** [comment]
In that case, you wouldn't need an adaptor, you could just:

```js
// get raw data, so links and computations are excluded
const data = ractive.get({ virtual: false });
// modify however you need to
data.items[2].a = 'something';
data.items.push({ a: 'sure' });
ractive.update();
```

Is there something else that the adaptor is managing?

**Comment by enneid** [comment]
this is unfortuantly more messy and complicated (maybe only in my mind :) ). This sample was only for reproducing bug.  
In my project I have isolated nested model, which I want to make sure, that, nothing external will change. It is independent from web solution (used also in node)  To use it in web I wrapped this model with decorator similar to backbone conception. This wrappers has list of attributes, which are or simple data , or next nested models. With that system I can modify then in many steps, and at the end  publish them to ractive. on only one level. If I will call on each nested level method ractive.set  this operation will be not atomic (many calls observers) and also it wold be killover, especially that that operations a are frequent.  For now to publish I used ractve.set(prefixer({changed_nested_model: {changed_more_data...} ...})), however i was was wonder if for simple change one key it wold't be better to even call  only ractive..update

**Comment by dagnelies** [comment]
Before this gets messy and unclear, I think an example would be appropriate. That way, be can measure actual performance bottlenecks and advise you on the best approach to take.

I too have doubts that adaptors/update are appropriate. ...premature optimization is the root of all evil as the saying goes.

By simply applying `Ractive.set('keypath', ...)` or `Ractive.set({'keypath1': ..., 'keypath2': ...})` you can usually update data fairly targeted. In my experience, performance is also mainly affected by the amount of nodes rendered and CSS rules ...much more so than the ractive code.

For example here: you can edit millions of items and it's super quick: http://dagnelies.github.io/ractive-dataset/big.html

**Comment by enneid** [comment]
Project, which i develop is now quite mature, and I think that changes in approach to adapters for now is not good idea. For me adaptors are strong features  becouse I can provide custom and computed data to ractive templates. From my perspective there are much more easy to use, than handmade ractive.set.
Besides project is running pretty well,  its  is WYSIWYG editor, with image, rich text ect... and there are now no major performing issues. 
I am aware that optimizing by lazy render of visible data is crucial, and there are some areas which can help me with speedup.

**Comment by evs-chris** [comment]
This should be fixed on edge as soon as travis finsihes travising. I'll push the fix out to the dev tags for 0.10 and 1.0 in a bit. Thanks for the report!
